File: cpp/src/store.cpp
#include "rohrpost/store.hpp"

#include "rohrpost/errors.hpp"
#include "rohrpost/io.hpp"
#include "rohrpost/paths.hpp"
#include "rohrpost/pyfmt.hpp"

#include <cerrno>
#include <cstring>
#include <format>

#if defined(_WIN32)
#include <fcntl.h>
#include <io.h>
#include <share.h>
#include <sys/locking.h>
#include <sys/stat.h>
#else
#include <fcntl.h>
#include <sys/file.h>
#include <sys/stat.h>
#include <unistd.h>
#endif

namespace rp::store {
namespace {

namespace fs = std::filesystem;

#if defined(_WIN32)
constexpr long kLockBytes = 1;  // lock and unlock must name the identical range

int open_lock_file(const fs::path& path) {
    int fd = -1;
    // "a+": read/write, create if missing, never truncate; binary so the CRT
    // does not touch bytes.
    _wsopen_s(&fd, path.c_str(), _O_RDWR | _O_CREAT | _O_APPEND | _O_BINARY, _SH_DENYNO, _S_IREAD | _S_IWRITE);
    return fd;
}
#else
int open_lock_file(const fs::path& path) {
    return ::open(path.c_str(), O_RDWR | O_CREAT | O_APPEND | O_CLOEXEC, 0644);
}
#endif

}  // namespace

FileLock::FileLock(const fs::path& rohrpost_dir) {
    const fs::path lock = paths::lock_path(rohrpost_dir);
    fd_ = open_lock_file(lock);
    if (fd_ < 0) {
        throw StoreError(std::format("cannot open {}: {}", io::path_str(lock), std::strerror(errno)));
    }
#if defined(_WIN32)
    // Byte range [0, 1) from the start of the file, exactly like msvcrt.locking
    // in the reference; LK_LOCK retries once a second for ~10 attempts.
    _lseek(fd_, 0, SEEK_SET);
    if (_locking(fd_, _LK_LOCK, kLockBytes) != 0) {
        const int err = errno;
        _close(fd_);
        fd_ = -1;
        throw StoreError(std::format(
            "could not lock {} within the ~10s wait budget (is another rp process holding it?): [Errno {}] {}",
            io::path_str(rohrpost_dir), err, std::strerror(err)));
    }
#else
    while (::flock(fd_, LOCK_EX) != 0) {
        if (errno == EINTR) continue;
        const int err = errno;
        ::close(fd_);
        fd_ = -1;
        throw StoreError(std::format("could not lock {}: {}", io::path_str(rohrpost_dir), std::strerror(err)));
    }
#endif
    locked_ = true;
}

FileLock::~FileLock() {
    if (fd_ < 0) return;
#if defined(_WIN32)
    if (locked_) {
        _lseek(fd_, 0, SEEK_SET);
        _locking(fd_, _LK_UNLCK, kLockBytes);
    }
    _close(fd_);
#else
    if (locked_) ::flock(fd_, LOCK_UN);
    ::close(fd_);
#endif
}

void append_event(const fs::path& rohrpost_dir, const Event& event) {
    const std::string line = encode(event) + "\n";
    FileLock lock(rohrpost_dir);
    const fs::path log = paths::log_path(rohrpost_dir);
#if defined(_WIN32)
    int fd = -1;
    _wsopen_s(&fd, log.c_str(), _O_WRONLY | _O_CREAT | _O_APPEND | _O_BINARY, _SH_DENYNO, _S_IREAD | _S_IWRITE);
    if (fd < 0) throw StoreError(std::format("cannot open {}: {}", io::path_str(log), std::strerror(errno)));
    const int written = _write(fd, line.data(), static_cast<unsigned>(line.size()));
    if (written < 0 || static_cast<std::size_t>(written) != line.size()) {
        const long long size = _filelengthi64(fd);
        const long long partial = written < 0 ? 0 : written;
        _chsize_s(fd, size - partial);
        _close(fd);
        throw StoreError(std::format(
            "short write to {}: wrote {} of {} bytes; appending the remainder would break single-write atomicity (§7)",
            io::path_str(log), partial, line.size()));
    }
    _close(fd);
#else
    const int fd = ::open(log.c_str(), O_WRONLY | O_CREAT | O_APPEND | O_CLOEXEC, 0644);
    if (fd < 0) throw StoreError(std::format("cannot open {}: {}", io::path_str(log), std::strerror(errno)));
    const ssize_t written = ::write(fd, line.data(), line.size());
    if (written < 0 || static_cast<std::size_t>(written) != line.size()) {
        // Roll back the partial bytes: a trailing half-line would otherwise fail
        // to decode on every future read (§3 principle 5).
        struct stat st{};
        const ssize_t partial = written < 0 ? 0 : written;
        if (::fstat(fd, &st) == 0) {
            const off_t truncated = st.st_size - static_cast<off_t>(partial);
            if (::ftruncate(fd, truncated) != 0) { /* best effort */ }
        }
        ::close(fd);
        throw StoreError(std::format(
            "short write to {}: wrote {} of {} bytes; appending the remainder would break single-write atomicity (§7)",
            io::path_str(log), partial, line.size()));
    }
    // os_sync: deliberately a no-op — git is the backup tier (see the reference).
    ::close(fd);
#endif
}

namespace {

/// Decode every line of every file (archive then log). Line numbers run
/// continuously across files, like the reference's chained iterator.
std::pair<std::vector<Event>, std::vector<std::string>> decode_all(const fs::path& rohrpost_dir) {
    std::vector<Event> events;
    std::vector<std::string> errors;
    std::size_t lineno = 0;
    std::vector<fs::path> files = paths::archive_files(rohrpost_dir);
    files.push_back(paths::log_path(rohrpost_dir));
    for (const auto& path : files) {
        std::error_code ec;
        if (!fs::is_regular_file(path, ec)) continue;
        auto content = io::read_file(path);
        if (!content) throw StoreError(std::format("cannot read {}: {}", io::path_str(path), content.error()));
        for (const auto raw : py::split_lines(*content)) {
            ++lineno;
            if (const auto bad = py::validate_utf8(raw)) {
                errors.push_back(std::format("line {}: {}", lineno, bad->message(raw)));
                continue;
            }
            const std::string_view stripped = py::strip(raw);
            if (stripped.empty()) continue;
            auto decoded = decode_line(stripped);
            if (decoded) {
                events.push_back(std::move(*decoded));
            } else {
                // `stripped[:80]!r` — the first 80 code points, Python-repr'd.
                std::size_t pos = 0;
                std::size_t count = 0;
                while (pos < stripped.size() && count < 80) {
                    if (!py::decode_utf8(stripped, pos)) ++pos;
                    ++count;
                }
                errors.push_back(std::format("line {}: {}: {}", lineno, decoded.error(), py::repr(stripped.substr(0, pos))));
            }
        }
    }
    return {std::move(events), std::move(errors)};
}

}  // namespace

std::vector<Event> read_events(const fs::path& rohrpost_dir) {
    auto [events, errors] = decode_all(rohrpost_dir);
    if (!errors.empty()) {
        throw StoreError(std::format("malformed event log ({} bad line(s)): {}", errors.size(), errors.front()));
    }
    return std::move(events);
}

std::pair<std::vector<Event>, std::vector<std::string>> read_events_lenient(const fs::path& rohrpost_dir) {
    return decode_all(rohrpost_dir);
}

std::size_t event_count(const fs::path& rohrpost_dir) {
    std::size_t count = 0;
    std::vector<fs::path> files = paths::archive_files(rohrpost_dir);
    files.push_back(paths::log_path(rohrpost_dir));
    for (const auto& path : files) {
        std::error_code ec;
        if (!fs::is_regular_file(path, ec)) continue;
        auto content = io::read_file(path);
        if (!content) continue;
        for (const auto raw : py::split_lines(*content)) {
            if (!py::strip(raw).empty()) ++count;
        }
    }
    return count;
}

}  // namespace rp::store

File: src/store.rs
//! The append-only event log: read, append, lock (spec §7).
//!
//! Every append happens under an exclusive lock on `.rohrpost/.lock` and is one
//! write of one line in append mode, so two writers never interleave
//! half-lines. The lock is the standard library's [`File::lock`]: `flock` on
//! Unix, `LockFileEx` on Windows — advisory, blocking, released when the holder
//! exits. Append mode is `O_APPEND` on Unix and `FILE_APPEND_DATA` on Windows.
//! A write that fails part-way is rolled back (the file is truncated to its
//! previous length) rather than resumed, so the log never keeps a corrupt tail.
//!
//! Reads are lock-free: the log is strictly append-only and every event carries
//! a unique id, so a partial final line (a truly concurrent writer) fails to
//! decode and is skipped by the lenient reader; the fold deduplicates.

use std::fs::{File, OpenOptions};
use std::io::Write as _;
use std::path::{Path, PathBuf};

use crate::error::{Error, Result, io_error};
use crate::events::{Event, decode_line};
use crate::paths;

/// An exclusive lock on `.rohrpost/.lock`, released on drop.
pub struct LockGuard {
    file: File,
}

impl Drop for LockGuard {
    fn drop(&mut self) {
        // Closing the file releases the lock too; the explicit unlock is hygiene.
        let _ = self.file.unlock();
    }
}

/// Take the exclusive store lock (blocks until the current holder releases).
/// Do not nest two calls on the same directory: the second blocks forever.
pub fn file_lock(rohrpost_dir: &Path) -> Result<LockGuard> {
    let path = paths::lock_path(rohrpost_dir);
    let file = OpenOptions::new()
        .read(true)
        .write(true)
        .create(true)
        .truncate(false)
        .open(&path)
        .map_err(|e| io_error("cannot open lock file", &path, &e))?;
    file.lock()
        .map_err(|e| io_error("cannot lock", &path, &e))?;
    Ok(LockGuard { file })
}

/// Append one event as a single JSONL line under the store lock.
pub fn append_event(rohrpost_dir: &Path, event: &Event) -> Result<()> {
    let mut line = event.encode();
    line.push('\n');
    let _guard = file_lock(rohrpost_dir)?;
    let log = paths::log_path(rohrpost_dir);
    let mut file = OpenOptions::new()
        .append(true)
        .create(true)
        .open(&log)
        .map_err(|e| io_error("cannot open", &log, &e))?;
    let before = file
        .metadata()
        .map_err(|e| io_error("cannot stat", &log, &e))?
        .len();
    if let Err(e) = file.write_all(line.as_bytes()) {
        // Roll back a partial line: a half-written tail would fail to decode on
        // every future read (spec §3 principle 5). Best-effort.
        let _ = file.set_len(before);
        return Err(io_error("short write to", &log, &e));
    }
    Ok(())
}

/// Archive files (oldest first) then the live log — the fold's read order (§6).
fn all_log_files(rohrpost_dir: &Path) -> Result<Vec<PathBuf>> {
    let mut files = paths::archive_files(rohrpost_dir)?;
    files.push(paths::log_path(rohrpost_dir));
    Ok(files)
}

/// Decode every line of every file, collecting malformed lines as messages.
fn decode_files(files: &[PathBuf]) -> Result<(Vec<Event>, Vec<String>)> {
    let mut events = Vec::new();
    let mut errors = Vec::new();
    for path in files {
        let bytes = match std::fs::read(path) {
            Ok(b) => b,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => continue,
            Err(e) => return Err(io_error("cannot read", path, &e)),
        };
        let name = path
            .file_name()
            .map(|n| n.to_string_lossy().into_owned())
            .unwrap_or_default();
        for (index, raw) in bytes.split(|&b| b == b'\n').enumerate() {
            let trimmed = trim_ascii(raw);
            if trimmed.is_empty() {
                continue;
            }
            let lineno = index + 1;
            match std::str::from_utf8(trimmed) {
                Err(_) => errors.push(format!("{name} line {lineno}: not valid UTF-8")),
                Ok(text) => match decode_line(text) {
                    Ok(event) => events.push(event),
                    Err(msg) => {
                        let preview: String = text.chars().take(80).collect();
                        errors.push(format!("{name} line {lineno}: {msg}: '{preview}'"));
                    }
                },
            }
        }
    }
    Ok((events, errors))
}

fn trim_ascii(bytes: &[u8]) -> &[u8] {
    let is_ws = |b: &u8| matches!(b, b' ' | b'\t' | b'\r' | b'\n');
    let start = bytes.iter().position(|b| !is_ws(b)).unwrap_or(bytes.len());
    let end = bytes
        .iter()
        .rposition(|b| !is_ws(b))
        .map_or(start, |i| i + 1);
    &bytes[start..end]
}

/// Every event from archive then log, in file order. Errors on the first
/// malformed line: a corrupt log is a loud failure. Duplicates are not removed
/// here; the fold deduplicates by event id.
pub fn read_events(rohrpost_dir: &Path) -> Result<Vec<Event>> {
    let (events, errors) = decode_files(&all_log_files(rohrpost_dir)?)?;
    if let Some(first) = errors.first() {
        return Err(Error::Store(format!(
            "malformed event log ({} bad line(s)): {first}",
            errors.len()
        )));
    }
    Ok(events)
}

/// Only the live `log.jsonl` (no archive), strictly decoded. Compaction moves
/// events out of this file and must never re-archive what is already archived.
pub fn read_live_events(rohrpost_dir: &Path) -> Result<Vec<Event>> {
    let (events, errors) = decode_files(&[paths::log_path(rohrpost_dir)])?;
    if let Some(first) = errors.first() {
        return Err(Error::Store(format!(
            "malformed event log ({} bad line(s)): {first}",
            errors.len()
        )));
    }
    Ok(events)
}

/// Like [`read_events`] but returns malformed lines instead of failing (for `rp doctor`).
pub fn read_events_lenient(rohrpost_dir: &Path) -> Result<(Vec<Event>, Vec<String>)> {
    decode_files(&all_log_files(rohrpost_dir)?)
}

/// Atomically replace `path` with `content` (write a sibling temp file, rename over).
pub fn write_atomic(path: &Path, content: &[u8]) -> Result<()> {
    let tmp = path.with_extension("tmp");
    std::fs::write(&tmp, content).map_err(|e| io_error("cannot write", &tmp, &e))?;
    std::fs::rename(&tmp, path).map_err(|e| io_error("cannot replace", path, &e))
}
